In [1]:
# import the library
%matplotlib inline

import pandas as pd
import numpy as np
import collections
import matplotlib.pyplot as plt
import seaborn as sns

from pyproj import Proj, transform

from sklearn.metrics.pairwise import euclidean_distances
from sklearn.cluster import DBSCAN

from sklearn import metrics

import bokeh
import bokeh.plotting as plotting
from bokeh.plotting import figure, show, ColumnDataSource
from bokeh.tile_providers import CARTODBPOSITRON
plotting.output_notebook()

sns.set_style('whitegrid')

# Date time
from datetime import datetime

# For distrance calculations between long and lat intersections
import math # from haversine import haversine
# Ref.:https://janakiev.com/blog/gps-points-distance-python/

Loading BokehJS ...

# Problem definition

Cluster regions based on exterminations

# Load the data

In [2]:
# Load the data
df_base = pd.read_csv('declarations-exterminations-punaises-de-lit-1.csv')

In [3]:
print(df_base.columns)
print('')
print(df_base.dtypes)

Index(['NO_DECLARATION', 'DATE_DECLARATION', 'DATE_PRIOR_INSP', 'EXT_FREQ',
       'DATE_FIRST_EXT', 'DATE_LAST_EXT', 'HOOD_NUM', 'HOOD_NAME', 'BORO_NAME',
       'MTM8_X', 'MTM8_Y', 'LONGITUDE', 'LATITUDE', 'LONG_LAT', 'MTM_X_Y',
       'DEC_MONTH', 'DEC_ISSUE', 'DATE_DIFF'],
      dtype='object')

NO_DECLARATION        int64
DATE_DECLARATION     object
DATE_PRIOR_INSP      object
EXT_FREQ            float64
DATE_FIRST_EXT       object
DATE_LAST_EXT        object
HOOD_NUM             object
HOOD_NAME            object
BORO_NAME            object
MTM8_X              float64
MTM8_Y              float64
LONGITUDE           float64
LATITUDE            float64
LONG_LAT             object
MTM_X_Y              object
DEC_MONTH             int64
DEC_ISSUE             int64
DATE_DIFF           float64
dtype: object


In [4]:
df_base['DATE_DECLARATION'] = pd.to_datetime(df_base['DATE_DECLARATION'])

In [5]:
# Inspication for epicenter calculation:
# https://towardsdatascience.com/transforming-categorical-data-for-usability-in-machine-learning-predictions-90459c3fc967?gi=2253c23cb822

# Ref. : https://stackoverflow.com/questions/4913349/haversine-formula-in-python-bearing-and-distance-between-two-gps-points
def haversine_np(lon1, lat1, lon2, lat2):
    lon1, lat1, lon2, lat2 = map(np.radians, [lon1, lat1, lon2, lat2])
    
    dlon = lon2 - lon1
    dlat = lat2 - lat1    
    
    a = np.sin(dlat / 2.0) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2.0) ** 2    
    c = 2 * np.arcsin(np.sqrt(a))
    
    km = 6367 * c
    return km

# Create distance column in dataframe, which returns km values
df_base['EPI_DIST_1'] = haversine_np(
    df_base['LONGITUDE'],df_base['LATITUDE'], -73.585636, 45.527404)

df_base['EPI_DIST_2'] = haversine_np(
    df_base['LONGITUDE'],df_base['LATITUDE'], -73.563652, 45.528809)

In [6]:
# Ref.:https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html
df_base['DATE_DECLARATION'] = pd.to_datetime(df_base['DATE_DECLARATION'])
df_base['DATE_PRIOR_INSP'] = pd.to_datetime(df_base['DATE_PRIOR_INSP'])
df_base['DATE_FIRST_EXT'] = pd.to_datetime(df_base['DATE_FIRST_EXT'])
df_base['DATE_LAST_EXT'] = pd.to_datetime(df_base['DATE_LAST_EXT'])

In [7]:
#We have this for month and year:

df_base['DEC_D'] = (pd.DatetimeIndex(df_base['DATE_DECLARATION']).year).map(str) + (
    pd.DatetimeIndex(df_base['DATE_DECLARATION']).month).map(str)
    
df_base['PRIOR_D'] = (pd.DatetimeIndex(df_base['DATE_PRIOR_INSP']).year).map(str) + (
    pd.DatetimeIndex(df_base['DATE_PRIOR_INSP']).month).map(str)

df_base['FIRST_D'] = (pd.DatetimeIndex(df_base['DATE_FIRST_EXT']).year).map(str) + (
    pd.DatetimeIndex(df_base['DATE_FIRST_EXT']).month).map(str)

df_base['LAST_D'] = (pd.DatetimeIndex(df_base['DATE_LAST_EXT']).year).map(str) + (
    pd.DatetimeIndex(df_base['DATE_LAST_EXT']).month).map(str)

# Create difference between LAST and FIRST Extermination
df_base['D_DIFF_B'] = round( (df_base['DATE_LAST_EXT'] - df_base['DATE_FIRST_EXT'] ) 
                            / np.timedelta64(1,'D') )
    # ['D_DIFF_B'] = ['DATE_LAST_EXT'] - ['DATE_FIRST_EXT']

# Create difference between LAST and FIRST Extermination
df_base['D_DIFF_C'] = round( (df_base['DATE_FIRST_EXT'] - df_base['DATE_PRIOR_INSP'] )
                            / np.timedelta64(1,'D') )
    # ['D_DIFF_C'] = ['DATE_FIRST_EXT'] - ['DATE_PRIOR_INSP']

    
# rename column
df_base['D_DIFF_A'] = df_base['DATE_DIFF'].copy() # ['D_DIFF_A'] = ['DATE_DECLARATION'] - ['DATE_PRIOR_INSP']

In [8]:
df_base.isnull().sum()
df_base.columns

Index(['NO_DECLARATION', 'DATE_DECLARATION', 'DATE_PRIOR_INSP', 'EXT_FREQ',
       'DATE_FIRST_EXT', 'DATE_LAST_EXT', 'HOOD_NUM', 'HOOD_NAME', 'BORO_NAME',
       'MTM8_X', 'MTM8_Y', 'LONGITUDE', 'LATITUDE', 'LONG_LAT', 'MTM_X_Y',
       'DEC_MONTH', 'DEC_ISSUE', 'DATE_DIFF', 'EPI_DIST_1', 'EPI_DIST_2',
       'DEC_D', 'PRIOR_D', 'FIRST_D', 'LAST_D', 'D_DIFF_B', 'D_DIFF_C',
       'D_DIFF_A'],
      dtype='object')

# Feature Engineering

In [9]:
# Ref.: https://github.com/arybressane/CEBD1260-BIG-DATA-ANALYTICS/blob/master/week7/clustering-dbscan-map.ipynb

# select a period
df = df_base[df_base['DATE_DECLARATION']>='2018-01-01']

# # remove lines with no location
df = df[(df['MTM8_X']>0)&(df['MTM8_Y']>0)]

# adapt X and Y to the visualization
df['MTM8_X'] = df.apply(lambda x: 
                        transform(Proj(init='epsg:4326'), Proj(init='epsg:3857'),
                                  x['LONGITUDE'], x['LATITUDE'])[1], axis=1)

df['MTM8_Y'] = df.apply(lambda x: 
                        transform(Proj(init='epsg:4326'), Proj(init='epsg:3857'),
                                  x['LONGITUDE'],x['LATITUDE'])[0], axis=1)

X_columns = ['MTM8_X', 'MTM8_Y'] # only two input columns are used

df = df[X_columns]

Only the 2018 period was considered for analysis, in an effort to improve run time.

If other features, other than longititude and latitude, are used, these other feature would need to be normalised and added into  separate columns. 

Specifically, if they are included as part of the X and Y coordinates (i.e. df['MTM8_X'] definition), the locations would be affected, which is not not intent. The result of the previous would create articificial locations, where it is possible that no extermination occured.

By having these features in a separate column, clusters can them be grouped by these attributes rather than the coordinates. Currently, clusters are grouped based on longitude and latitude coordinates.

In [10]:
df

MTM8_X        MTM8_Y
388    5.705930e+06 -8.188897e+06
389    5.707264e+06 -8.187637e+06
390    5.709142e+06 -8.193115e+06
391    5.701257e+06 -8.195114e+06
392    5.707259e+06 -8.191682e+06
393    5.705930e+06 -8.188897e+06
394    5.706389e+06 -8.188774e+06
395    5.706583e+06 -8.192794e+06
404    5.719110e+06 -8.194897e+06
405    5.699442e+06 -8.190936e+06
407    5.707415e+06 -8.203596e+06
409    5.707324e+06 -8.186342e+06
411    5.689215e+06 -8.193773e+06
413    5.697613e+06 -8.195720e+06
414    5.699442e+06 -8.190936e+06
415    5.693255e+06 -8.197389e+06
416    5.708568e+06 -8.193670e+06
420    5.706779e+06 -8.204008e+06
423    5.705370e+06 -8.206314e+06
424    5.702359e+06 -8.189543e+06
427    5.703534e+06 -8.189621e+06
428    5.697613e+06 -8.195720e+06
429    5.689032e+06 -8.192964e+06
431    5.712931e+06 -8.186459e+06
432    5.712625e+06 -8.193897e+06
433    5.713307e+06 -8.195067e+06
434    5.707264e+06 -8.187637e+06
435    5.699797e+06 -8.192192e+06
436    5.706355e+06 -8.189924e+06
439    5.697344e+06 -8.191746e+06
...             ...           ...
30938  5.704642e+06 -8.192070e+06
30956  5.706677e+06 -8.192205e+06
30957  5.710308e+06 -8.187192e+06
30958  5.713745e+06 -8.189785e+06
31003  5.713745e+06 -8.189785e+06
31004  5.702984e+06 -8.190203e+06
31005  5.710308e+06 -8.187192e+06
31021  5.711948e+06 -8.193407e+06
31022  5.701741e+06 -8.194808e+06
31044  5.701741e+06 -8.194808e+06
31045  5.708763e+06 -8.193009e+06
31046  5.701988e+06 -8.190242e+06
31047  5.706367e+06 -8.192934e+06
31070  5.696321e+06 -8.189781e+06
31071  5.696908e+06 -8.191780e+06
31072  5.699468e+06 -8.190421e+06
31073  5.713745e+06 -8.189785e+06
31074  5.707324e+06 -8.186342e+06
31100  5.703695e+06 -8.201546e+06
31101  5.704174e+06 -8.192035e+06
31120  5.702463e+06 -8.190503e+06
31138  5.702189e+06 -8.188710e+06
31161  5.688971e+06 -8.195897e+06
31179  5.705943e+06 -8.201912e+06
31180  5.701741e+06 -8.194808e+06
31194  5.709026e+06 -8.194798e+06
31209  5.701741e+06 -8.194808e+06
31211  5.706915e+06 -8.186820e+06
31212  5.693375e+06 -8.189527e+06
31213  5.701741e+06 -8.194808e+06

[3628 rows x 2 columns]

# Model Training

In [11]:
model = DBSCAN(eps=3.0, min_samples=4)
model.fit(df[['MTM8_X', 'MTM8_Y']])

# Ref.: https://stackoverflow.com/questions/26666367/scikit-dbscan-eps-and-min-sample-value-determination
# Number of clusters in labels, ignoring noise if present.
db=model.fit(df[['MTM8_X', 'MTM8_Y']])
labels = db.labels_
n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)
print('Estimated number of clusters: %d' % n_clusters_)


cluster_labels = model.labels_
n_clusters = len(set(cluster_labels))
print(collections.Counter(cluster_labels))

df['cluster'] = cluster_labels

Estimated number of clusters: 237
Counter({-1: 1773, 43: 60, 7: 49, 28: 39, 4: 37, 5: 31, 3: 28, 32: 23, 84: 23, 79: 22, 9: 21, 74: 21, 50: 20, 72: 20, 85: 19, 122: 18, 129: 16, 1: 15, 14: 15, 15: 15, 63: 15, 106: 15, 234: 15, 26: 14, 53: 14, 60: 14, 64: 14, 114: 13, 120: 13, 8: 12, 10: 12, 13: 12, 17: 12, 45: 12, 66: 12, 130: 12, 21: 11, 41: 11, 47: 11, 90: 11, 133: 11, 18: 10, 55: 10, 112: 10, 119: 10, 136: 10, 140: 10, 180: 10, 181: 10, 51: 9, 73: 9, 105: 9, 108: 9, 131: 9, 153: 9, 200: 9, 215: 9, 220: 9, 30: 8, 39: 8, 44: 8, 56: 8, 83: 8, 87: 8, 101: 8, 110: 8, 123: 8, 124: 8, 150: 8, 164: 8, 170: 8, 176: 8, 179: 8, 191: 8, 192: 8, 193: 8, 211: 8, 16: 7, 19: 7, 25: 7, 27: 7, 37: 7, 42: 7, 68: 7, 76: 7, 89: 7, 98: 7, 113: 7, 137: 7, 166: 7, 199: 7, 236: 7, 0: 6, 2: 6, 23: 6, 31: 6, 36: 6, 40: 6, 52: 6, 58: 6, 61: 6, 67: 6, 69: 6, 77: 6, 86: 6, 91: 6, 92: 6, 93: 6, 94: 6, 111: 6, 116: 6, 127: 6, 154: 6, 162: 6, 163: 6, 183: 6, 190: 6, 204: 6, 207: 6, 223: 6, 226: 6, 6: 5, 12: 5, 29: 

In [12]:
df

MTM8_X        MTM8_Y  cluster
388    5.705930e+06 -8.188897e+06        0
389    5.707264e+06 -8.187637e+06        1
390    5.709142e+06 -8.193115e+06        2
391    5.701257e+06 -8.195114e+06       -1
392    5.707259e+06 -8.191682e+06       -1
393    5.705930e+06 -8.188897e+06        0
394    5.706389e+06 -8.188774e+06       -1
395    5.706583e+06 -8.192794e+06        3
404    5.719110e+06 -8.194897e+06       -1
405    5.699442e+06 -8.190936e+06        4
407    5.707415e+06 -8.203596e+06        5
409    5.707324e+06 -8.186342e+06        6
411    5.689215e+06 -8.193773e+06        7
413    5.697613e+06 -8.195720e+06        8
414    5.699442e+06 -8.190936e+06        4
415    5.693255e+06 -8.197389e+06        9
416    5.708568e+06 -8.193670e+06       10
420    5.706779e+06 -8.204008e+06       11
423    5.705370e+06 -8.206314e+06       -1
424    5.702359e+06 -8.189543e+06       -1
427    5.703534e+06 -8.189621e+06       -1
428    5.697613e+06 -8.195720e+06        8
429    5.689032e+06 -8.192964e+06       12
431    5.712931e+06 -8.186459e+06       13
432    5.712625e+06 -8.193897e+06       -1
433    5.713307e+06 -8.195067e+06       -1
434    5.707264e+06 -8.187637e+06        1
435    5.699797e+06 -8.192192e+06       -1
436    5.706355e+06 -8.189924e+06       14
439    5.697344e+06 -8.191746e+06       -1
...             ...           ...      ...
30938  5.704642e+06 -8.192070e+06      215
30956  5.706677e+06 -8.192205e+06       -1
30957  5.710308e+06 -8.187192e+06       -1
30958  5.713745e+06 -8.189785e+06      236
31003  5.713745e+06 -8.189785e+06      236
31004  5.702984e+06 -8.190203e+06       -1
31005  5.710308e+06 -8.187192e+06       -1
31021  5.711948e+06 -8.193407e+06       -1
31022  5.701741e+06 -8.194808e+06      234
31044  5.701741e+06 -8.194808e+06      234
31045  5.708763e+06 -8.193009e+06       -1
31046  5.701988e+06 -8.190242e+06       -1
31047  5.706367e+06 -8.192934e+06      108
31070  5.696321e+06 -8.189781e+06       -1
31071  5.696908e+06 -8.191780e+06       -1
31072  5.699468e+06 -8.190421e+06       -1
31073  5.713745e+06 -8.189785e+06      236
31074  5.707324e+06 -8.186342e+06        6
31100  5.703695e+06 -8.201546e+06       -1
31101  5.704174e+06 -8.192035e+06       -1
31120  5.702463e+06 -8.190503e+06      223
31138  5.702189e+06 -8.188710e+06      235
31161  5.688971e+06 -8.195897e+06       -1
31179  5.705943e+06 -8.201912e+06       -1
31180  5.701741e+06 -8.194808e+06      234
31194  5.709026e+06 -8.194798e+06       -1
31209  5.701741e+06 -8.194808e+06      234
31211  5.706915e+06 -8.186820e+06       -1
31212  5.693375e+06 -8.189527e+06      192
31213  5.701741e+06 -8.194808e+06      234

[3628 rows x 3 columns]

In [13]:
p = figure(y_range=(5641788.0, 5751788.0), x_range=(-8152883, -8252883))
p.add_tile(CARTODBPOSITRON)

latitude  = list(df[df['cluster']>-1]['MTM8_X'].values)
longitude = list(df[df['cluster']>-1]['MTM8_Y'].values)

colormap = list(bokeh.palettes.viridis(n_clusters))
colors = [colormap[x] for x in df[df['cluster']>-1]['cluster']]
source = ColumnDataSource(data=dict(longitude=longitude, latitude=latitude))
p.circle(x=longitude, y=latitude, color=colors, fill_alpha=0.2, size=5)
show(p)

# Model Evaluation

In [14]:
# Inter-Cluster
centroids = []
for cluster in sorted(set(model.labels_)):
    centroids.append(df[df['cluster']==cluster][X_columns].mean().values)
distances = []
for c1 in centroids:
    for c2 in centroids:
        distances.append(euclidean_distances(c1.reshape(-1, 1), c2.reshape(-1, 1))[0][0])
print('Inter Cluster distance', np.mean(distances))

# Intra-Cluster
distances = []
for cluster in sorted(set(model.labels_)):
    df_filter = df[df['cluster']==cluster]
    centroid = df_filter[X_columns].mean().values
    for k, v in df_filter[X_columns].iterrows():
        distances.append(euclidean_distances(centroid.reshape(-1, 1), v.values.reshape(-1, 1))[0][0])
print('Intra Cluster distance', np.mean(distances))

# Inertia
distances = []
for cluster in sorted(set(model.labels_)):
    df_filter = df[df['cluster']==cluster]
    centroid = df_filter[X_columns].mean().values
    for k, v in df_filter[X_columns].iterrows():
        distances.append(euclidean_distances(centroid.reshape(1, -1), v.values.reshape(1, -1), squared=True)[0][0])
print('Inertia', np.sum(distances))



Inter Cluster distance 7774.905569554047
Intra Cluster distance 2686.464515078813
Inertia 145331572692.98438


In [15]:
# Ref.: https://medium.com/@elutins/dbscan-what-is-it-when-to-use-it-how-to-use-it-8bd506293818
print("Silhouette Coef: %0.3f" % metrics.silhouette_score(df[['MTM8_X', 'MTM8_Y']],labels))

# "A silhouette score ranges from -1 to 1, with -1 being the worst score possible and 1 being the best score." from Ref.
# "Silhouette scores of 0 suggest overlapping clusters." from Ref.

Silhouette Coef: 0.051


# Describe clusters

###### Describe your clusters (write a description/summary for each one of the clusters on the notebook)

By increasing epsilon, the amount of noise is increased, and by decreasing the min_samples in tandem, the silhouette score improve from a negative float value to a positive value near 0. The improvide value of 0.051 was obtained with model = DBSCAN(eps=3.0, min_samples=4). Highier values of min_samples resulted in negative values of Silhouette coefficient.

The result only considers the longitude and latitude for the inputs, without any consideration for the distance from epicenter and for the dates relative to the intersection. These variable could have been introduced and normalized into the model; thereafter, a filter could have been applied, in order to review only the exertinations within a radius or distance from an epicenter in the Plateau neighborhood as an example.

